database/src/backend/heed/database.rs

```diff
@@ -204,79 +168,48 @@ impl<T: Table> DatabaseRo<T> for HeedTableRo<'_, T> {
 impl<T: Table> DatabaseRo<T> for HeedTableRw<'_, '_, T> {
     #[inline]
     fn get(&self, key: &T::Key) -> Result<T::Value, RuntimeError> {
-        get::<T>(&self.db, self.tx_rw, key)
-    }
-
-    #[inline]
-    fn get_range<'a, Range>(
-        &'a self,
-        range: Range,
-    ) -> Result<impl Iterator<Item = Result<T::Value, RuntimeError>> + 'a, RuntimeError>
-    where
-        Range: RangeBounds<T::Key> + 'a,
-    {
-        get_range::<T, Range>(&self.db, self.tx_rw, range)
-    }
-
-    #[inline]
-    fn iter(
-        &self,
-    ) -> Result<impl Iterator<Item = Result<(T::Key, T::Value), RuntimeError>> + '_, RuntimeError>
-    {
-        iter::<T>(&self.db, self.tx_rw)
-    }
-
-    #[inline]
-    fn keys(
-        &self,
-    ) -> Result<impl Iterator<Item = Result<T::Key, RuntimeError>> + '_, RuntimeError> {
-        keys::<T>(&self.db, self.tx_rw)
-    }
-
-    #[inline]
-    fn values(
-        &self,
-    ) -> Result<impl Iterator<Item = Result<T::Value, RuntimeError>> + '_, RuntimeError> {
-        values::<T>(&self.db, self.tx_rw)
+        get::<T>(&self.db, &self.tx_rw.borrow(), key)
     }
 
     #[inline]
     fn len(&self) -> Result<u64, RuntimeError> {
-        len::<T>(&self.db, self.tx_rw)
+        len::<T>(&self.db, &self.tx_rw.borrow())
     }
 
     #[inline]
     fn first(&self) -> Result<(T::Key, T::Value), RuntimeError> {
-        first::<T>(&self.db, self.tx_rw)
+        first::<T>(&self.db, &self.tx_rw.borrow())
     }
 
     #[inline]
     fn last(&self) -> Result<(T::Key, T::Value), RuntimeError> {
-        last::<T>(&self.db, self.tx_rw)
+        last::<T>(&self.db, &self.tx_rw.borrow())
     }
 
     #[inline]
     fn is_empty(&self) -> Result<bool, RuntimeError> {
-        is_empty::<T>(&self.db, self.tx_rw)
+        is_empty::<T>(&self.db, &self.tx_rw.borrow())
     }
 }
 
 impl<T: Table> DatabaseRw<T> for HeedTableRw<'_, '_, T> {
     #[inline]
     fn put(&mut self, key: &T::Key, value: &T::Value) -> Result<(), RuntimeError> {
-        Ok(self.db.put(self.tx_rw, key, value)?)
+        Ok(self.db.put(&mut self.tx_rw.borrow_mut(), key, value)?)
     }
 
     #[inline]
     fn delete(&mut self, key: &T::Key) -> Result<(), RuntimeError> {
-        self.db.delete(self.tx_rw, key)?;
+        self.db.delete(&mut self.tx_rw.borrow_mut(), key)?;
         Ok(())
     }
 
     #[inline]
     fn pop_first(&mut self) -> Result<(T::Key, T::Value), RuntimeError> {
```

Comment:
Just for confirmation, `DatabaseRw::pop_{first,last}()` are still safe, correct?

We have `&mut heed::RwTxn` and aren't returning any value with a lifetime.

Reply:
yeah they are ok